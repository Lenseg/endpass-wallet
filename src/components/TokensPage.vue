<template lang="html">
  <nav class="panel">
  <p class="panel-heading">
    Tokens
  </p>
  <div class="panel-block">
    <p class="control has-icons-left">
      <input v-model="search" class="input is-small" type="text" placeholder="search">
      <span class="icon is-small is-left">
        <i class="fas fa-search" aria-hidden="true"></i>
      </span>
    </p>
  </div>
  <!-- <p class="panel-tabs">
    <a class="is-active">all</a>
    <a>public</a>
    <a>private</a>
    <a>sources</a>
    <a>forks</a>
  </p> -->
  <div class="scroller">
    <a v-for="token in filteredTokens" class="panel-block is-clearfix is-block">
      <span class="panel-icon">
        <i class="fas fa-book" aria-hidden="true"></i>
      </span>
      {{token.symbol}}
      <span class="is-pulled-right" v-if="token.balance">{{token.balance}}</span>
      <button v-else @click="saveToken(token)" class="button is-pulled-right" type="button" name="button">Show</button>
    </a>
  </div>
  <div class="panel-block">
    <button class="button is-link is-outlined is-fullwidth">
      reset all filters
    </button>
  </div>
</nav>
</template>

<script>
export default {
  data() {
    return {
      tokens: [],
      search: ''
    }
  },
  computed: {
    filteredTokens() {
      if(this.search === '')
        return this.tokens
      else
        return this.tokens.filter((token) => {
          return token.symbol.includes(this.search) || token.name.includes(this.search)
        });
    }
  },
  created() {
    this.$store.dispatch('tokens/createSubscribtion').then(()=> {
      this.$store.state.tokens.subscription.on('update', this.setBalances)
    });
    this.$store.dispatch('tokens/getTokens')
      .then(response => {
      this.tokens = response.body.map((token)=> {
        token.balance = null;
        return token
      });
      this.setBalances();
    });
  },
  methods: {
    saveToken(token) {
      this.$store.dispatch('tokens/subscribeOnTokenUpdates', token.address, token);
      this.setBalances();
    },
    setBalances() {
      this.tokens.forEach((token)=>{
        let balance = null;
        console.log(this.$store.state.tokens.activeTokens)
        let newToken = this.$store.state.tokens.activeTokens.find((balance) => {
          if(balance.address === token.address)
            token.balance = balance.balance;
        });
      })
    }
  }
}
</script>

<style lang="css">
.scroller {
  max-height: 500px;
  overflow: scroll;
}
</style>
