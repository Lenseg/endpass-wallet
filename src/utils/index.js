export const kebabToCamel = input =>
  input.replace(/-([a-z])/g, g => g[1].toUpperCase());
