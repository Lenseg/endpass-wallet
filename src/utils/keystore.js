import keythereum from 'keythereum';
import bs58check from 'bs58check';
import { kdfParams } from '@/config';

// Monkey patch keythereum to skip generating address for private keys
// This allows us to encrypt private keys of arbitrary length, and
// conforms better to the Ethereum keystore V3 spec, which does not include
// the address for privacy reasons.
// See https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition#alterations-from-version-1
keythereum.privateKeyToAddress = function(pk) {
  return '';
};

export default {
  // Encrypts a private key Buffer into a V3 keystore object
  // The exported keystore does NOT include an address
  encrypt(password, privateKey) {
    // Generate random salt and iv for each encryption
    let dk = keythereum.create();
    let options = {
      kdf: kdfParams.kdf,
      kdfparams: kdfParams,
    };
    let encrypted = keythereum.dump(
      password,
      privateKey,
      dk.salt,
      dk.iv,
      options,
    );
    delete encrypted.address;
    return encrypted;
  },

  // Decrypts a V3 keystore object into a private key Buffer
  decrypt(password, keystore) {
    return keythereum.recover(password, keystore);
  },

  // Encode to Base58Check
  encodeBase58(key) {
    return bs58check.encode(key);
  },

  // Decode from Base58Check
  decodeBase58(key) {
    return bs58check.decode(key);
  },
};
