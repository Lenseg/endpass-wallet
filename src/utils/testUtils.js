const toKebab = string =>
  string.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let dashName;

    if (stubComponent.name) {
      dashName = toKebab(stubComponent.name);
    }

    stubs[dashName] = {
      render(createElement) {
        return createElement(dashName, this.$slots.default);
      },
    };

    return stubs;
  }, {});
}
