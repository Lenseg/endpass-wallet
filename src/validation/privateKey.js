import ethUtils from 'ethereumjs-util';
export default {
  getMessage(field, params, data) {
    return (data && data.message) || 'Something went wrong';
  },
  validate(value, args) {
    let isKey;
    try {
      isKey = ethUtils.isValidPrivate(
        Buffer.from(value.replace(/^0x/, ''), 'hex'),
      );
    } catch (e) {
      return {
        valid: false,
        data: { message: e.message },
      };
    }
    return {
      valid: isKey,
      data: !isKey ? { message: 'This is not a valid private key' } : undefined,
    };
  },
};
