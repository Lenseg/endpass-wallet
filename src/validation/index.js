import Vue from 'vue';
import { default as VeeValidate, Validator } from 'vee-validate';
import address from './address';
import privateKey from './privateKey';
import publicKey from './publicKey';
import seedPhrase from './seedPhrase';
import hex from './hex';

Validator.extend('address', address);
Validator.extend('private_key', privateKey);
Validator.extend('public_key', publicKey);
Validator.extend('seed_phrase', seedPhrase);
Validator.extend('hex', hex);

Vue.use(VeeValidate);
