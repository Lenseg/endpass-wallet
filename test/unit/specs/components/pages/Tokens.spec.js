import { shallow, createLocalVue, mount } from '@vue/test-utils';
import Vuex from 'vuex';
import Notifications from 'vue-notification';
import validation from '@/validation';

import TokensPage from '@/components/pages/Tokens.vue';
import { generateStubs } from '@/utils/testUtils';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(Notifications);

describe('TokensPage', () => {
  let wrapper;
  let actions;
  let getters;
  let store;
  let options;
  let tokens;

  beforeEach(() => {
    tokens = [
      {
        name: 'First Token',
        symbol: 'FST',
        address: '0x4ce2109f8db1190cd44bc6554e35642214fbe144',
      },
      {
        name: 'second token',
        symbol: '$SCdT',
        address: '0x4bd5c3e7e4d6b3df23e9da5b42e5e4daa3d2579b',
      },
    ];

    actions = {
      updateTokenPrice: jest.fn(),
      addTokenToSubscription: jest.fn(),
      removeTokenFromSubscription: jest.fn(),
      getAllTokens: jest.fn(() => tokens),
    };

    getters = {
      savedActiveTokens: () => [{}],
      net: () => 1,
    };

    store = new Vuex.Store({
      modules: {
        accounts: {
          namespaced: true,
          state: {
            settings: {
              fiatCurrency: 'USD',
            },
          },
        },
        tokens: {
          namespaced: true,
          state: {
            activeTokens: tokens,
            prices: null,
          },
          actions,
          getters,
        },
        price: {
          namespaced: true,
          state: {
            price: null,
          },
        },
      },
    });

    options = {
      store,
      localVue,
    };
  });

  afterEach(() => {
    actions.updateTokenPrice.mockClear();
    actions.addTokenToSubscription.mockClear();
    actions.removeTokenFromSubscription.mockClear();
    actions.getAllTokens.mockClear();
  });

  describe('render', () => {
    beforeEach(() => {
      wrapper = shallow(TokensPage, {
        ...options,
        stubs: generateStubs(TokensPage),
      });
    });

    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('tokens-page');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('v-spinner', () => {
      // it('should render v-spinner', () => {
      //   wrapper.setComputed({
      //     activeTokens: [],
      //   });
      //
      //   expect(wrapper.find('v-spinner').attributes()).toEqual({
      //     'is-loading': 'true',
      //     class: 'spinner-block',
      //   });
      // });

      it('should not render v-spinner', () => {
        wrapper.setComputed({
          activeTokens: [{}, {}],
        });

        expect(wrapper.find('v-spinner').attributes()).toEqual({
          class: 'spinner-block',
        });
      });
    });
  });

  describe('behavior', () => {
    beforeEach(() => {
      wrapper = shallow(TokensPage, options);
    });

    describe('search', () => {
      it("should correctly find the user's tokens", () => {
        expect(wrapper.vm.userTokenList).toHaveLength(2);

        wrapper.setData({
          search: 'fst',
        });

        expect(wrapper.vm.userTokenList).toHaveLength(1);
      });

      it('should correctly find tokens in list', async () => {
        expect(wrapper.vm.searchTokenList).toHaveLength(2);

        wrapper.setData({
          tokens: [
            ...tokens,
            {
              name: 'Third Token',
              symbol: 'TTKN',
              address: '0x687422eea2cb73b5d3e242ba5456b782919afc85',
            },
            {
              name: 'fours token',
              symbol: 'FurT',
              address: '0xab54de61a908583e6332a1282c7bfca39f899b4f',
            },
          ],
        });

        expect(wrapper.vm.$data.tokens).toHaveLength(4);

        wrapper.setData({
          searchToken: 'ttk',
        });

        expect(wrapper.vm.searchTokenList).toHaveLength(1);

        wrapper.setData({
          searchToken: '',
        });

        expect(wrapper.vm.searchTokenList).toHaveLength(4);

        wrapper.setData({
          searchToken: 'hir',
        });

        expect(wrapper.vm.searchTokenList).toHaveLength(1);
      });
    });

    describe('user token', () => {
      it('should call remove action', () => {
        const button = wrapper.find('span#remove-token-1');
        button.trigger('click');

        expect(actions.removeTokenFromSubscription).toHaveBeenCalledTimes(1);
        expect(actions.removeTokenFromSubscription).toBeCalledWith(
          expect.any(Object),
          tokens[1],
          undefined,
        );
      });
    });
  });

  describe('props', () => {
    beforeEach(() => {
      wrapper = shallow(TokensPage, options);
    });

    it('should call an action when changing the net id', () => {
      expect(wrapper.vm.userTokenList).toHaveLength(2);

      wrapper.setComputed({
        net: 3,
      });

      expect(actions.getAllTokens).toHaveBeenCalledTimes(2);
    });
  });
});
