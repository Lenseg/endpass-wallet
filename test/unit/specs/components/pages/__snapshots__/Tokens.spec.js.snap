// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokensPage render should render component 1`] = `
<div
  class="tokens-page"
>
  <div
    class="section"
  >
    <div
      class="container"
    >
      <div
        class="columns"
      >
        <div
          class="column is-half"
        >
          <div
            class="card app-card"
          >
            <div
              class="card-content"
            >
              <nav
                class="panel"
              >
                <p
                  class="panel-heading"
                >
                  
                  Your Tokens
                  
                  <button
                    class="button is-primary"
                    name="button"
                  >
                    <span
                      class="icon panel-icon is-small"
                    >
                      {}
                    </span>
                  </button>
                </p>
                 
                <div
                  class="panel-block"
                >
                  <search-input />
                </div>
                 
                <v-spinner
                  class="spinner-block"
                />
                 
                <div
                  class="scroller"
                >
                  <a
                    class="panel-block is-clearfix is-block"
                  >
                    <span
                      class="token-symbol"
                    >
                      FST
                    </span>
                     
                    <span
                      class="token-name"
                    >
                      First Token
                    </span>
                     
                    <balance
                      amount="NaN"
                      currency=""
                    />
                     
                    <!---->
                     
                    <span
                      class="icon has-text-danger is-medium"
                      id="remove-token-0"
                    >
                      {}
                    </span>
                  </a>
                  <a
                    class="panel-block is-clearfix is-block"
                  >
                    <span
                      class="token-symbol"
                    >
                      $SCdT
                    </span>
                     
                    <span
                      class="token-name"
                    >
                      second token
                    </span>
                     
                    <balance
                      amount="NaN"
                      currency=""
                    />
                     
                    <!---->
                     
                    <span
                      class="icon has-text-danger is-medium"
                      id="remove-token-1"
                    >
                      {}
                    </span>
                  </a>
                </div>
              </nav>
            </div>
          </div>
        </div>
         
        <div
          class="column is-half"
        >
          <h4
            class="title is-4"
          >
            Add token
          </h4>
           
          <!---->
        </div>
      </div>
    </div>
  </div>
   
  <!---->
</div>
`;
